### PR TITLE
soc: riscv: telink_b91: add dfu support for Telink b91 platform

### PR DESCRIPTION
**Pull request by rikorsev**
This PR contains functionality required for correct work of Telink b91 platform with DFU subsystem and **mcuboot** bootloader.
It contains following changes:
- updates in default Kconfig and device tree for **tlsr9518adk80d** board
- fixes in UART driver to make it works properly after booting from **mcuboot** bootloader.
- add config for **smp_srv** sample.
- changes in linker script and Kconfig files for b91 soc.

Signed-off-by: Alexandr Kolosov <rikorsev@gmail.com>

### CODE REVIEW

File: samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay
Comment (lairdjm) [inline samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay:20]: Why are these being added to the smp_svr sample? The main UART should be part of the board dts not a specific application overlay

File: drivers/flash/soc_flash_b91.c
Comment (edersondisouza) [inline drivers/flash/soc_flash_b91.c:108]: typo: "erace"

File: samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay
Comment (rikorsev) [reply samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay]: Hi @lairdjm 

Thanks for remark. I've been removed uart0 from overlay but uart1 is still there, due to it is not part of dts for tlsr9518adk80d board. Shall I move uart1 to tlsr9518adk80d board as well?

Pretty the same approach as I've been used I saw here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts

Regards,
Alex

File: drivers/flash/soc_flash_b91.c
Comment (rikorsev) [reply drivers/flash/soc_flash_b91.c]: Thanks! Fixed.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
Flash driver extension is topic of other PR already https://github.com/zephyrproject-rtos/zephyr/pull/45940
Changes to different modules should go in dedicated commits. Tangled ones might go in one commit.

File: west.yml
Comment (nvlsianpu) [inline west.yml:177]: such reference might be used only for testing purpose. We are using zephyr-rtos/mcuboot fork here.
I will make its update right after the 3.1.0 release.

File: samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay
Comment (lairdjm) [reply samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay]: It needs to be part of the board DTS yes, you cannot add an overlay for this board to every single sample application, that is untenable and a major maintenance burden

Comment (rikorsev) [reply samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay]: @lairdjm 
Ok, got it. The .overlay currently removed. I guess dts stuff would be maintained by separate PR.

File: west.yml
Comment (rikorsev) [reply west.yml]: @nvlsianpu 

Ok, thanks for info. Update to west.yml has been removed from PR.

**Comment by rikorsev** [comment]
@nvlsianpu 

Thanks for review. I've been refactor it according to the remarks. Flash related stuff has been removed from the PR.
Rest code has been divided into smaller commits depends on touched module. 

Regards, 
Alex

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts
Comment (fkokosinski) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:119]: Could you make this comment 80-cols wide?

**Review by fkokosinski — COMMENTED** [review]
Minor nitpick, looks good otherwise.

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts
Comment (rikorsev) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts]: Done

**Review by edersondisouza — DISMISSED** [review]

File: samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf
Comment (lairdjm) [inline samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf:1]: I can't access the datasheet for this part, is there any particular reason that this needs to be set to 256 instead of the default 512?

Comment (rikorsev) [reply samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf:1]: Hi @lairdjm 

Thanks for nice question. The answer is: b91 platform has 256 bytes flash page size.

Just for curiosity, I've check it once again and remove the config (which means that IMG_BLOCK_BUF_SIZE has default value 512). **mcumgr** image update operation hung immediately after start. The reason is that function **find_flash_total_size** (stream_flash.c) return false on the condition `if (ctx->buf_len > info->size)`. Also I see log message "Buffer size is bigger than page"

**Review by fkokosinski — DISMISSED** [review]

File: samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf
Comment (lairdjm) [reply samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf:1]: @de-nordic Should this go in the boards folder in Kconfig.defconfig with an if around it or in the mcumgr sample? Seems like it would be applicable to many samples and end applications

**Review by fkokosinski — DISMISSED** [review]

**Comment by rikorsev** [comment]
Hi @edersondisouza, @lairdjm, @nvlsianpu,

Please review the PR and approve it if there is no remarks from your side.

Regards,
Alex

**Comment by lairdjm** [comment]
@de-nordic Can you check the comment in the sample application overlay?

File: samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf
Comment (rikorsev) [reply samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf]: Hi @lairdjm, @de-nordic 

The config is moved into tlsr9518adk80d_defconfig with detailed comment. Please, take a look and approve the PR if everything is ok.

Regards,
Alex

**Review by edersondisouza — CHANGES_REQUESTED** [review]
I guess that with this change you should simply remove the "[samples: smp_svr: add config for tlsr9518adk80d](https://github.com/zephyrproject-rtos/zephyr/pull/45973/commits/fa3202333f6a326423599c13a5845b60666d8acb)" patch. No need to add a file in one patch to just remove it in the next.
Please, make this `CONFIG_IMG_BLOCK_BUF_SIZE` a single patch on the board default config.

**Comment by rikorsev** [comment]
Hi @edersondisouza 

Nice catch! You are completely right. Redundant patch has been removed.

Regards,
Alex

**Review by edersondisouza — APPROVED** [review]

**Comment by lairdjm** [comment]
@rikorsev this is no longer mcumgr related, please stop adding myself and de-nordic

**Comment by rikorsev** [comment]
Hi @kgugala, @pgielda, @dcpleung, @kasunch, @mgielda, @tgorochowik, @nvlsianpu, fkokosinski

Please, review the PR and approve it if it is fine.

Regards,
Alex

**Review by dcpleung — APPROVED** [review]

**Review by tgorochowik — APPROVED** [review]

**Comment by rikorsev** [comment]
Hi @nvlsianpu,

Could you close the **change request**, as it already satisfied? It is blocking merge of the PR.

Regards,
Alex

**Review by nvlsianpu — APPROVED** [review]

**Pull request merged**
